cpp.mk

```diff
@@ -0,0 +1,181 @@
+
+$${BUILD_DIR}/$1/coverage.xml : $${BUILD_DIR}/$1/.$1_already_run
+	@../modules/gcovr/scripts/gcovr $${$1_COVERAGE_ARGS} --xml > $$@
+
+.PHONY : coverage_check_$1
+coverage_check_$1 : $${BUILD_DIR}/$1/coverage.xml
+	@mkdir -p $${BUILD_DIR}/scratch/
+	@xmllint --xpath '//class[@line-rate!="1.0"]/@filename' $$< \
+		| tr ' ' '\n' \
+		| grep filename= \
+		| cut -d\" -f2 \
+		| sort > $${BUILD_DIR}/scratch/coverage_$1.tmp
+	@sort ut/coverage-not-yet | comm -23 $${BUILD_DIR}/scratch/coverage_$1.tmp - > $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp
+	@if grep -q ^ $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp ; then \
+		echo "Error: some files unexpectedly have less than 100% code coverage:" ; \
+		cat $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp ; \
+		exit 2 ; \
+	fi
+CLEANS += $${BUILD_DIR}/scratch/coverage_$1.tmp $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp
+
+.PHONY : coverage_raw_$1
+coverage_raw_$1 : $${BUILD_DIR}/$1/.$1_already_run
+	@../modules/gcovr/scripts/gcovr $${$1_COVERAGE_ARGS} --keep --sort-percentage
+
+test : run_$1
+valgrind : valgrind_$1
+vg_check : valgrind_check_$1
+coverage_check : coverage_check_$1
+coverage_raw : coverage_raw_$1
+
+CLEANS += $${BUILD_DIR}/$1/valgrind_output.xml $${BUILD_DIR}/$1/.$1_already_run
+
+endef
+
+# Default values for build flags for each build
+__COMMON_CPPFLAGS := -ggdb3 -std=c++11
+release_CPPFLAGS := -O2 ${__COMMON_CPPFLAGS}
+test_CPPFLAGS := -O0 ${__COMMON_CPPFLAGS} -DUNIT_TEST \
+                 -fprofile-arcs -ftest-coverage -fno-access-control\
+								 -I${GTEST_DIR}/include -I${GMOCK_DIR}/include \
+                 -I${CPP_COMMON_DIR}/test_utils
+test_LDFLAGS := -lgcov
+
+ifdef JUSTTEST
+  EXTRA_TEST_ARGS ?= --gtest_filter=*$(JUSTTEST)*
+endif
+
+../build/obj/gmock-all.o : ${GMOCK_DIR}/src/gmock-all.cc ${GMOCK_DIR}/include/gmock/*.h ${GMOCK_DIR}/include/gmock/internal/*.h
+	@mkdir -p ../build/obj
+	${CXX} ${test_CPPFLAGS} -I${GTEST_DIR}/include -I${GMOCK_DIR}/include -I${GMOCK_DIR} -c $< -o $@
+../build/obj/gtest-all.o : ${GTEST_DIR}/src/gtest-all.cc ${GTEST_DIR}/include/gtest/*.h ${GTEST_DIR}/include/gtest/internal/*.h
+	@mkdir -p ../build/obj
+	${CXX} ${test_CPPFLAGS} -I${GTEST_DIR}/include -I${GTEST_DIR}/include -I${GTEST_DIR} -c $< -o $@
+
+# Print out the generate Makefile snippet for debugging purposes
+ifdef DEBUG_MAKEFILE
+$(foreach target,${TARGETS},$(info $(call common_target,${target},release)))
+$(foreach target,${TEST_TARGETS},$(info $(call test_target,${target})))
+$(error Not building since DEBUG_MAKEFILE was specified)
+endif
+
+# Expand the build definitions for each provided target
+$(foreach target,${TARGETS},$(eval $(call common_target,${target},release)))
+$(foreach target,${TEST_TARGETS},$(eval $(call test_target,${target})))
+
+# Build all non-test targets by default
+all : ${TARGETS}
+
+clean :
+	-rm $(sort ${CLEANS}) # make's sort function removes duplicates as a side effect
+
+print-% :
```

Comment:
Ah, dark Makefile magic:

```
$ make print-VARIABLE
VARIABLE = <value of VARIABLE>
```

Handy for debugging, I think it's in some (but not all) of our `platform.mk` files.